### ISSUE TITLE
Badge: Import Users from a file to the DB using the front-end

### ISSUE THREAD

**Issue by LucasBrazi06** [sap-labs-france/ev-dashboard#2203]
1. Upload the file in the DB per 1Mb => Json format (only Admin)
  - [x] Check how the browser is handling the big file in memory
2. Upload the file in the MongoDB special collection files
  - [ ] Check with Jerome if we can save it chunk by chunk
  - [x] Load the file chunk by chunk
  - [x] A chuck is a valid JSon document
  - [x] Alternative: Store each user to import in a dedicated collection: 1 user -> 1 Json document in DB
3. For each chunk => Put a lock to prevent 2 user imports in //
  - [ ] For each user
  - [x] Apply a JSON schema
  - [x] Import the user
    - [ ] Check the repository: User, Tags, Site ID
    - [x] Imported on/by
  - [x] Status = Success
    - [x] Mark the user as successfully imported in the DB to not process it again
    - [ ] Or remove it?
   - [x] Status = Failure
     - [x] Provide the reason of the failure
4. Display a message when it's finished using socket.IO to the end user
  - [ ] Only the user who trigger the import
5. Build an UI that will display the users that are in the importusers collection
  - [ ] List the user's data + status of the import (progression)
  - [ ] Trigger again the import of all users in the list or only one user
  - [ ] Delete selected users (faulted or not)
  - [ ] Provide import users button
6. Document it in the readme.md

**Comment by LucasBrazi06** [comment]
Also to take into account the import of the badge only.
Scenario:
- The CPO orders 10k badges and import them in the DB.
- These badges are not related to a user so free to be assigned later on.
- Users with role basic receive their badges and declare them in their profile (other issue tbd)

**Comment by LucasBrazi06** [comment]
@OliveGerste You will be in charge of the implementation
@ClaudeROSSI You will work with Tobias on tech choices and PR review with keeping me posted offline
Thanks to both of you

**Comment by fcoutellier** [comment]
FYI, God requested for 15 feb latest.

**Comment by LucasBrazi06** [comment]
@mahdibenromdhane Is there checkboxes to tick at the beginning of the issue? ;-)

**Comment by LucasBrazi06** [comment]
@ClaudeROSSI Could you check the 2 PRs asap?

**Comment by mahdibenromdhane** [comment]
> @mahdibenromdhane Is there checkboxes to tick at the beginning of the issue? ;-)

done

**Comment by ClaudeROSSI** [comment]
ooops ... removed high priority by mistake

**Comment by LucasBrazi06** [comment]
@mahdibenromdhane I put in back in the In Progress pipeline

**Comment by ClaudeROSSI** [comment]
@mahdibenromdhane: As already discussed, I did a sync with Serge. 
Here is the outcome:

General comment is that we somehow have more time to implement it. Proviridis may have other options to actually do the import of their new users. So, we take this additional time to go further with the first implementation!

1) User Activation
Serge would like that users  get imported in a pending state (because email verification is a prerequisite to the user account activation).
For this, we need to clarify what happens if the a pending user (with no password) tries to reset its password. May be some adaptations are then necessary to make it possible.
Once the user resets its password, he/she will then be in a regular situation. The normal flow for the email verification and the account activation should occur without any additional effort.

2) Two consecutive uploads
The data of the second upload should override the data of the previous one. The second upload should also be able to remove some data (for instance, if the CSV file has less columns than the initial one)

3) Concurrent Uploads
This is not a use-case that we need to support. However, to be on the safe side,  it might be necessary to avoid that situation. Can we put a lock while doing the upload? The concurrent upload may receive a feedback that the operation can not be performed, and suggest to try it later.

4) Concurrent Synch of Users
This should not happen because the actual job doing it will be triggered by import. i.e.: Two imports == two jobs.
Here we assume that we will have for each upload a new entry added in a kind of "todo" list and a particular task (executed every minute) will check the todo list for jobs to execute.
In other words, each user synchronisation will be triggered separately and the locking mechanism will prevent from running two user sync in parallel.

Let me know if you need more details.
Thanks

**Comment by ClaudeROSSI** [comment]
@LucasBrazi06 -Import of Tags server side is ready.
Same status than the User Import. No feedback is sent when the import is done.
